Throttle scan I/O when other I/Os are in progress

mattmacy opened the pull request:
Signed-off-by: Matt Macy <mmacy@FreeBSD.org>

<!--- Please fill out the following template, which will help other contributors review your Pull Request. -->

<!--- Provide a general summary of your changes in the Title above -->

<!---
Documentation on ZFS Buildbot options can be found at
https://github.com/zfsonlinux/zfs/wiki/Buildbot-Options
-->

### Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

### Description
<!--- Describe your changes in detail -->

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
<!--- If your change is a performance enhancement, please provide benchmarks here. -->

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

mattmacy commented:
@kithrup Does ZoL need this? They've dropped tick based throttling.

mattmacy commented:
> @kithrup Does ZoL need this? They've dropped tick based throttling.

@kithrup Brian doesn't believe that this is still useful. I'm going to revert in ZoF unless you have reason to believe otherwise.

snajpa commented:
Yes please!

I've just seen IOs being held in the queue for as long as 34 seconds (zpool iostat -w reporting) during scrub, paused it, all good afterwards.

snajpa commented on module/zfs/dsl_scan.c:
```diff
@@ -137,6 +137,16 @@ extern int zfs_vdev_async_write_active_min_dirty_percent;
+ * Tune with care. Very high settings (hundreds) are known to trigger
+ * some firmware bugs and resets on certain SSDs.
+ */
+int zfs_top_maxinflight = 32;		/* maximum I/Os per top-level */
```
Um, I don't see this one being actually used anywhere...

mattmacy replied:
Looks like I missed the actual use of it:
```
/* ARGSUSED */
static void
dsl_scan_prefetch_thread(void *arg)
{
	dsl_scan_t *scn = arg;
	spa_t *spa = scn->scn_dp->dp_spa;
	vdev_t *rvd = spa->spa_root_vdev;
	uint64_t maxinflight = rvd->vdev_children * zfs_top_maxinflight;
	scan_prefetch_issue_ctx_t *spic;

	/* loop until we are told to stop */
	while (!scn->scn_prefetch_stop) {
		arc_flags_t flags = ARC_FLAG_NOWAIT |
                    ARC_FLAG_PRESCIENT_PREFETCH | ARC_FLAG_PREFETCH;
		int zio_flags = ZIO_FLAG_CANFAIL | ZIO_FLAG_SCAN_THREAD;
		
		mutex_enter(&spa->spa_scrub_lock);

		/*
		 * Wait until we have an IO to issue and are not above our
		 * maximum in flight limit.
		 */
		while (!scn->scn_prefetch_stop &&
		    (avl_numnodes(&scn->scn_prefetch_queue) == 0 ||
		    spa->spa_scrub_inflight >= scn->scn_maxinflight_bytes)) {
			cv_wait(&spa->spa_scrub_io_cv, &spa->spa_scrub_lock);
		}

```

mattmacy replied:
I don't know if these sorts of arbitrary limits make sense. They don't tend to age well. However, I'm not sure how easy it would be to implement some sort of PID controller for it.

snajpa replied:
PID controllers need careful tuning, we'd need to come up with some zfs-tuned or something, I guess...

richardelling replied:
zfs_top_maxinflight did not work well in practice, mostly because it was multiplied by the number of children, so hundreds was not uncommon.  In any case, this throttling mechanism proved to be nearly impossible to optimize in the field.

The current ZIO scheduler approach seems to work better, though the defaults are questionable and biased against scan workloads.

behlendorf commented:
To add a little more context to this, the previous throttle was pretty simplistic and dropped intentionally in the sequential resilver work.  The updated code relies on the existing vdev queue throttling mechanisms.  Adding some kind of throttle back would be reasonable if we can show it's useful, @snajpa's comment suggests that's the case.  Though, I'd strongly prefer we use hrtime for it instead of ticks for portability between platforms.  How long a tick is varies wildly.

mattmacy commented:
> To add a little more context to this, the previous throttle was pretty simplistic and dropped intentionally in the sequential resilver work. The updated code relies on the existing vdev queue throttling mechanisms. Adding some kind of throttle back would be reasonable if we can show it's useful, @snajpa's comment suggests that's the case. Though, I'd strongly prefer we use hrtime for it instead of ticks for portability between platforms. How long a tick is varies wildly.

By default, on server hardware, HZ=1000 on FreeBSD (Embedded and VMs use HZ=100). So we could just translates the numbers to milliseconds if we're interested in going forward with this.

mattmacy commented:
> To add a little more context to this, the previous throttle was pretty simplistic and dropped intentionally in the sequential resilver work. The updated code relies on the existing vdev queue throttling mechanisms. Adding some kind of throttle back would be reasonable if we can show it's useful, @snajpa's comment suggests that's the case. Though, I'd strongly prefer we use hrtime for it instead of ticks for portability between platforms. How long a tick is varies wildly.

@behlendorf I've updated the change to use hrtime.

behlendorf commented on module/zfs/zio.c:
```diff
@@ -3526,7 +3526,25 @@ zio_vdev_io_start(zio_t *zio)
+		uint64_t new = NSEC2MSEC(gethrtime());
+		if (old != new)
+			(void) atomic_cas_64(&spa->spa_last_io, old, new);
+	}
```
This logic all predates the addition of the new write throttle and vdev queuing code.  Rather that simply reinstating the previous logic, I'm sure we can do something a little better now.  That would be desirable since if I recall correctly this throttle mechanism was difficult to tune in practice and frequently resulted in sub-optimal scrub and resilver times since it could starve out scrub/resilver I/Os.

One possibly worth exploring would be to update `vdev_queue_class_to_issue()` to not issue pending scrub I/O when there are outstanding or pending synchronous writes or reads.  I'd expect that moving the throttle itself closer to the device would improve our overall disk utilization.  Furthermore, you might not want to throttle for non-rotation devices and this is something we could take in to account at this lower layer.

@tcaputi and @amotin may have some additional thoughts about this.

tcaputi replied:
I'm interested to see some performance numbers / pathological cases here. Back when I was working on this code, we tested it fairly extensively (since the whole point was to make resilvers and scrubs more predictable and faster when there was headroom for it). I agree that the old mechanism had a good number of issues, which was why we removed it when we rewrote the scanning code.

I'm particularly interested in the example @snajpa was talking about. Holding IO's in a queue for 34 seconds definitely seems like a bug.

mattmacy replied:
@behlendorf My only concern at the moment is working through the diff between FreeBSD and master. Should I close this PR and revert the change in ZoF, trusting that any underlying issues will be dealt with elsewhere?

behlendorf replied:
I think before we make any change in this area we need to understand and characterize the performance implications of that change.  I don't mind leaving this open as a WIP if someone wants to use this as a starting point for that investigation.

behlendorf commented:
Closing PR for now.  This is an issue we intend to revisit after the initial OpenZFS refactoring is complete.
